[RFC] Ephemeral resources and write-only attributes

yottta opened the pull request:
<!-- If your PR resolves an issue, please add it here. -->
Part of #1996
[Rendered view](https://github.com/opentofu/opentofu/blob/ephemeral-rfc/rfc/20250317-ephemeral-resources.md)

## Target Release

1.11.0

## Checklist

<!-- Please check of ALL items in this list for all PRs: -->

- [ ] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [ ] I have not used an AI coding assistant to create this PR.
- [ ] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.
- [ ] I (and other contributors to this PR) have not looked at the Terraform source code while implementing this PR.

### Go checklist

<!-- If your PR contains Go code, please make sure you check off all items on this list: --> 

- [ ] I have run golangci-lint on my change and receive no errors relevant to my code.
- [ ] I have run existing tests to ensure my code doesn't break anything.
- [ ] I have added tests for all relevant use cases of my code, and those tests are passing.
- [ ] I have only exported functions, variables and structs that should be used from other packages.
- [ ] I have added meaningful comments to all exported functions, variables, and structs.

### Website/documentation checklist

<!-- If you have changed the website, please follow this checklist: -->

- [ ] I have locally started the website as [described here](https://github.com/opentofu/opentofu/blob/main/website/README.md) and checked my changes.

Yantrio commented on rfc/20250317-ephemeral-resources.md:
`````diff
@@ -0,0 +1,646 @@
+  ````
+* If a module is having an ephemeral variable declared, that variable can get values from any source, even another non-ephemeral variable.
+
+We should use boolean marks, as no additional information is required to be carried. When introducing the marks for these, extra care should be taken in *all* the places marks are handled and ensure that the existing implementation around marks is not affected.
`````
do we also need to introduce some checks/tests to ensure that functions such as `sensitive()` do not accidentally strip the ephemeral mark too?

yottta replied:
3fe9247caedeef0099138111dece7a18065cda2b

Yantrio commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+> For an example on how to properly introduce a new mark in the outputs, you can check the [PR](https://github.com/opentofu/opentofu/pull/2633) for the deprecated outputs.
+
+Strict rules:
+* A root module cannot define ephemeral outputs since are useless.
```
```suggestion
* A root module cannot define ephemeral outputs.
```

yottta replied:
ce97918f1ad6429e17d88f7c250530742926f3f0

Yantrio commented on rfc/20250317-ephemeral-resources.md:
````diff
@@ -0,0 +1,646 @@
+}
+```
+
+Once a local is marked as ephemeral, this can be used only in other ephemeral contexts. Check the `User Documentation` section for more details on the allowed contexts.
+
````
let's not forget to document here that variables that take in ephemeral marked values should also be implicitly marked as ephemeral

yottta replied:
Added information in 11a29cb050fd95ab55b5f6d5df01bba3c73af854.
A variable will not become ephemeral strictly from referencing an ephemeral value. In order for it to be able to work with ephemeral values, it needs to be configured specifically.

Yantrio commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+* variables
+* outputs
+* locals
+* `ephemeral` resources
```
I just want to go on record that I don't like how they're called `ephemeral resources` because this is confusing. Mainly because `resource` and `ephemeral` here are 2 different types of blocks.

I would prefer a different name, such as "ephemeral entities" etc, but I fear it may be too late.

yottta replied:
I don't think that it's too late and I agree that it can become confusing. If others agree with this, it's an easy change:
* "%s/ephemeral resource/ephemeral entity/g"
* "%s/ephemeral resources/ephemeral entities/g"

Hitting enter? 😅

Yantrio commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+
+### Ephemeral resources
+Due to the fact ephemeral resources are not stored in the state/plan file, this block is not creating a diff in the OpenTofu's UI.
+Instead, OpenTofu will notify the user of opening/renewing/closing an ephemeral resource with messages similar to the following:
```
Can we discuss what these actions mean? `opening` `renewing` and `closing`?  As someone approaching this fresh I think I can assume what it means but I think we should declare it explicitly so we can have a conversation about it. 

It feels like a strange set of verbs to show up in the ui to me (Opening.. Closing... etc)

yottta replied:
Changed a little in 24b57e7c7b16816ae8e5d98b57658781656179c9.
But please check the specific chapter dedicated to the `Renew` [method](https://github.com/opentofu/opentofu/blob/6b81cb76c65cb76311db9ecf9d67a94fbb0185cc/rfc/20250317-ephemeral-resources.md#renew-method-details).

Yantrio commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+* Also, both blocks support `Configure`, `ConfigValidators` and `ValidateConfig` as extensions of the basic definition.
+* Ephemeral resources do support two more operations in contrast with datasources:
+  * `Renew`
+    * Together with the data returned by the `Open` method call, the provider can also specify a `RenewAt` which will be a specific moment in time when OpenTofu should call the `Renew` method to get an updated information from the ephemeral resource. OpenTofu will have to check for `RenewAt` value anytime it intends to use the value returned by the ephemeral resource.
```
This renewat functionality isn't super clear to me based on the doc above, I wasn't expecting this. 

Can you describe it in abit more detail and why it's needed? Is it so that providers can determine if/when opentofu should refetch rather than it only being a human (version) driven re-fetch?

yottta replied:
Is the `Renew` [method] (https://github.com/opentofu/opentofu/blob/6b81cb76c65cb76311db9ecf9d67a94fbb0185cc/rfc/20250317-ephemeral-resources.md#renew-method-details) dedicated section enough?

Yantrio commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+  * `Renew`
+    * Together with the data returned by the `Open` method call, the provider can also specify a `RenewAt` which will be a specific moment in time when OpenTofu should call the `Renew` method to get an updated information from the ephemeral resource. OpenTofu will have to check for `RenewAt` value anytime it intends to use the value returned by the ephemeral resource.
+  * `Close`
+    * When an ephemeral resource is having this method defined, it is expecting it to be called in order to release a possible held resource. A good example of this is with a Vault provider that could provide a secret by obtaining a lease, and when the secret is done being used, OpenTofu should call `Close` on that ephemeral resource to instruct on releasing the lease and revoking the secret.
```
```suggestion
    * When an ephemeral resource is having this method defined, it is expecting it to be called in order to release a possible held resource. A good example of this is with a Vault/OpenBao provider that could provide a secret by obtaining a lease, and when the secret is done being used, OpenTofu should call `Close` on that ephemeral resource to instruct on releasing the lease and revoking the secret.
```

yottta replied:
da777bb65a5a2162f48c80bca72301d6412b937f

Yantrio commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+  * `Renew`
+    * Together with the data returned by the `Open` method call, the provider can also specify a `RenewAt` which will be a specific moment in time when OpenTofu should call the `Renew` method to get an updated information from the ephemeral resource. OpenTofu will have to check for `RenewAt` value anytime it intends to use the value returned by the ephemeral resource.
+  * `Close`
+    * When an ephemeral resource is having this method defined, it is expecting it to be called in order to release a possible held resource. A good example of this is with a Vault provider that could provide a secret by obtaining a lease, and when the secret is done being used, OpenTofu should call `Close` on that ephemeral resource to instruct on releasing the lease and revoking the secret.
```
>it to be called in order to release a possible held resource.

When, how?

yottta replied:
Added in https://github.com/opentofu/opentofu/commit/24b57e7c7b16816ae8e5d98b57658781656179c9#diff-1b793b731c0b3e0929632ddfa87429e5d0d2ddbe95bd3f9a09dea465ac1d406dR587-R597

Do you think that we should go more in-depth? Most probably the work of this should be done in `graphNodeCloseProvider`, but I think that this kind of details should be left for the implementor to decide.
Please correct me if I am wrong here.

Yantrio commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+
+> [!NOTE]
+>
+> If any information in the configuration of an ephemeral resource is unknown during the `plan` phase, OpenTofu will defer the provisioning of the resource for the `apply` phase.
```
```suggestion
> If any information in the configuration of an ephemeral resource is unknown during the `plan` phase, OpenTofu will defer the provisioning of the resource for the `apply` phase. This means that inconsistency can occur between the plan and the apply phase.
```

yottta replied:
f603db886aa3d830534b5cc68f7e415e58e3d343

Yantrio commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+> [!NOTE]
+> 
+> `Renew` does not return a *new* information meant to replace the initial `Result` returned by the `Open` call.
+> Due to this, `Renew` is only useful for system similar to Vault where the lease can be renewed without generating new data.
```
```suggestion
> Due to this, `Renew` is only useful for systems where an entity can be renewed without generating new data.
```

yottta replied:
f603db886aa3d830534b5cc68f7e415e58e3d343

Yantrio commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+> Due to this, `Renew` is only useful for system similar to Vault where the lease can be renewed without generating new data.
+
+#### `Close` method details
+When OpenTofu is done using an ephemeral resource, it needs to call its `Close` method to ensure that any remote data associated with the data returned in `OpenResponse.Result` is released and/or cleaned up properly.
```
When you say "Done using", is that
- when the apply phase is finished
- when the value is no longer referenced by a resource/output?
- when the node in the graph has been read completely

Can you elaborate on what `done` is here please?

yottta replied:
Yeah, I am having a rough time wording that correctly. "done using" in this context means that the apply phase finished and now it executes the close nodes. Right before executing `graphNodeCloseProvider` nodes, it should call `Close` on the ephemeral resources.
If you can help with a good wording around this, would greatly appreciate it.

Yantrio commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+
+#### `Close` method details
+When OpenTofu is done using an ephemeral resource, it needs to call its `Close` method to ensure that any remote data associated with the data returned in `OpenResponse.Result` is released and/or cleaned up properly.
+The `Renew` request is requiring the latest `Private` data returned by the call to `Open` or `Renew` method.
```
I think this line needs to be further up, not under the `Close` method details

Yantrio commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+There is not much to say here, since this is the same lifecycle that a datasource is having.
+
+#### Testing support
+The testing support will be documented later into a different RFC, or as amendment to this one.
```
Can you give a reason as to why we're not putting it in this RFC?

yottta replied:
I didn't dig into this part yet and I barely touched the testing feature of this project.
If you think that this should be added, I can try my best but might take some time until I get a good understanding of the things in there.

cam72cam replied:
Given the limited usage of the testing feature, I agree with @yottta here.  We should focus on the main impact of this feature.

Yantrio replied:
I think I'm just looking for some reasoning to be added to the RFC. I'm fine with not doing it now, we should just document why.

yottta replied:
Quick fix: dd4ca4729bfb730c406cd06ceaa52e2d6fcfea25

Yantrio commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+
+### Utilities
+#### `terraform.applying`
+The `terraform.applying` needs to be introduced to allow the user to check if the current command that is running is `apply` or not.
```
why terraform and not tofu?

yottta replied:
Because of the feature parity. Maybe having both just to be able to migrate? And in case we encounter `terraform.applying` write a warning that we also support `tofu.applying`? 🤔

cam72cam replied:
Internally, they both map to the same structure for now.  Let's at least mention tofu.applying here.

yottta replied:
Done e8c189287f194d9ea8f2319658605ad22d618f45

Yantrio commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+
+> [!NOTE]
+>
+> This keyword is related to the `apply` command and not to the `apply` phase, meaning that when running `tofu apply`, `terraform.applying` will still be `true` also during the `plan` phase of the `apply` command.
```
what about during destroy?

yottta replied:
Fixed in 2c6587907eae9c404931ce789091dfd192f5628b

Yantrio commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+This is an ephemeral value that should be handled correctly and ensure that its value or any other value generate from it will not end up in a plan/state file.
+
+#### `ephemeralasnull` function
+`ephemeralsnull` function is useful when an object that was built by referencing an ephemeral value wants to be used into a non-ephemeral context.
```
should we recommend that people use this with a nice warning in the UI when they try and create root module outputs with ephemeral values ?

yottta replied:
That's a good point 🤔 I like it! Thanks
Added in 2c6587907eae9c404931ce789091dfd192f5628b

cam72cam commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+
+Issue: https://github.com/opentofu/opentofu/issues/1996
+
+Right now, OpenTofu information for resources and outputs are written to state as it is. This is presenting a security risk
+as some of the information from the stored objects can contain sensitive bits that can become visible to whoever is having access to the state file.
```
It also might be worth introducing the idea that state encryption is our currently recommended solution.

IMO they are two overlapping, but not exclusive concepts

Advantages of State Encryption:
* All secrets are protected, regardless of provider or resource/module configuration
* Don't need to chose what is protected and what is exposed
* Also protects against provider injection
  - https://www.plerion.com/blog/hacking-terraform-state-for-privilege-escalation
  
  Advantages of Ephemeral
  * No keys to loose / have leaked, secrets (if identified properly) are not stored in the first place.

Yantrio replied:
That raises another question, we should explain how we shouldn't use this with state encryption (due to the ephemerality) because some people may think "This is great for secrets, i can use it for my passphrase" and have issues.

yottta replied:
Please check this out: ae712fdaf7a5ba4b69cfcf0e65d960df7d301dac

cam72cam commented:
Overall, this is a really good start!  I think the main area I'd like to see expanded is the user documentation.  Your example is quite comprehensive and also quite dense.  I think introducing a few simpler examples, before going with the "Here's the a complex example showing how the simpler building blocks interplay" would be helpful.

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+a non-write-only argument called `secret_version`. Every time the user wants to update the value of `secret`, it needs to change the value of `secret_version` to trigger a change.
+The provider implementation is responsible with handling this particular case: because the version attribute is stored also in the state, the provider needs to compare the value from the state with the one from the configuration and in case it differs, it will trigger the update of the `secret` attribute.
+
+The write-only attributes are supported momentarily by a low number of providers and resources.
```
In 91f105241ac49c3c20ba51d52306be62c7ac4c03

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+Right now, OpenTofu information for resources and outputs are written to state as it is. This is presenting a security risk
+as some of the information from the stored objects can contain sensitive bits that can become visible to whoever is having access to the state file.
+
+In order to provide a better solution for the aforementioned situation, OpenTofu introduces the concept of "ephemerality".
```
Please check this out ae712fdaf7a5ba4b69cfcf0e65d960df7d301dac

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+* connection blocks
+* provider configuration
+
+Usage in any other place will raise an error:
```
Changed in 119b5c9a3c32f5a41a671586e4852515b4ee3b34. Indeed there were a lot of places. Sorry for this, seems a bias on my side. Weird.

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+* `connection` and `provisioner` blocks
+
+### Ephemeral resource
+In contrast with the write-only arguments where only specifically tagged attributes are skipped from the state/plan file, `ephemeral` resources are skipped entirely.
```
e7c684a0fbb6f68fd87bcaeec74e17c7978d76bb

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+
+As seen in this particular change of the [terraform-plugin-framework](https://github.com/hashicorp/terraform-plugin-framework/commit/ecd80f67daed0b92b243ae59bb1ee2077f8077c7), the write-only attribute cannot be configured for set attributes, set nested attributes and set nested blocks.
+
+Write-only attributes cannot generate a plan diff because the prior state does not contain a value that OpenTofu can use to compare the new value against and also the planned value of a write-only argument will always be empty.
```
b4121ce87a6c902a783827039b5f38ddac6d7bc6

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+OpenTofu will not store ephemeral variable(s) in plan files. 
+If a plan is generated from a configuration that is having at least one ephemeral variable, 
+when the planfile will be applied, the value(s) for the ephemeral variable(s) needs to passed again by 
+using `-var` or `-var-file` arguments.
```
c47bbb084a80d34a729715b39653b14bc870f3e5

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+not from the root module. 
+This limitation is natural since ephemeral outputs are meant to be skipped from the state file. Therefore, there is no usage of such a defined output block in a root module.
+
+Ephemeral outputs are useful when a child module returns sensitive data, allowing the caller to use the value of that output in other ephemeral contexts.
```
14bc104ef981f19fe02640e00ee25e76ccab2070

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+### Outputs
+An `output` block can be configured as ephemeral as long as it's
+not from the root module. 
+This limitation is natural since ephemeral outputs are meant to be skipped from the state file. Therefore, there is no usage of such a defined output block in a root module.
```
603979cec2bf7e3c76064ef7515982f3c5d1c703

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+- (1) Variables will be able to be marked as ephemeral. By doing so, those will be able to be used only in ephemeral contexts.
+- (2) Version field that is going together with the actual write-only argument to be able to update the value of it. To upgrade the secret, the version field needs to be updated, otherwise OpenTofu will generate no diff for it.
+- (3) Using ephemeral resource to retrieve the secret. Maybe looks a little bit weird, because right above we are having the resource of the same type that is looking like it should be able to be used to get the secret. In reality, because that `resource` is using `secret_string_wo` to store the information, that field is going to be null when referenced. Check (9) for more details.
+- (4) Module output that is referencing an ephemeral value, it needs to be marked as ephemeral too. Otherwise, OpenTofu will generate an error.
```
4d02f23b7bca34f689e30afa7290ed9cdd355c7d

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+### Write-only arguments
+Most of the write-only arguments logic is already in the [provider-framework](https://github.com/hashicorp/terraform-plugin-framework):
+* [Initial implementation](https://github.com/hashicorp/terraform-plugin-framework/pull/1044)
+* [Sets comparisson enhancement](https://github.com/hashicorp/terraform-plugin-framework/pull/1064)
```
11730c4361e2d3137791ceeb79a6a3cce5444d44

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+- (5) The variable that is going to be used in an ephemeral variable, is not required to be ephemeral. The value can also be a hardcoded value without being ephemeral.
+- (6) Here we used another `aws_secretsmanager_secret` just to have an easier example on how ephemeral/write-only/(ephemeral variables)/(ephemeral outputs) are working together. But there will be more and more resources that will allow to work with this new concept.
+- (7) Referencing a module ephemeral output to ensure that the ephemeral information is passed correctly between two modules.
+- (8) Creating an output that is referencing a write-only argument *requires* the output to be marked as sensitive. 
```
Changed here: 88a7826941e9048742cd732405ac2cf6f3417da6
I agree that would make more sense on having this also for the child modules, but do we want to create this diff between TF and OpenTofu?

Yantrio replied:
I think we should consider if it's useful or not. I don't have much of an opinion either way here really. I feel like root level only is fine.

yottta replied:
Follow up on this: After further investigation on this one, this was having **nothing** to do with write-only argument being referenced in an output. The provider field I was using, was marked as `sensitive` in the provider schema. Therefore, [this](https://github.com/opentofu/opentofu/blob/d1f0999aed2b171f62ae802c23b9674ca0afda6c/internal/tofu/node_output.go#L377-L389) is the code that was executed.
Still, the rule with only the root output still stands but it's in OpenTofu for a long time now.

yottta commented on rfc/20250317-ephemeral-resources.md:
`````diff
@@ -0,0 +1,646 @@
+  ```hcl
+  var.password (ephemeral)
+     Enter a value:
+  ````
`````
Thanks. Fixed in 4d02f23b7bca34f689e30afa7290ed9cdd355c7d

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,646 @@
+
+For enabling ephemeral variables, these are the basic steps that need to be taken:
+* Update config to support the `ephemeral` attribute.
+* Mark the variables with a new mark ensure that the marks are propagated correctly.
```
I am not aware of this. Could you expand it a little bit please?

Yantrio replied:
https://github.com/opentofu/opentofu/pull/2503

yottta replied:
Oh, this is what you meant! Ok. I misread your initial comment.
Updated in 2695d6c58dc739ceae192242733ddea987cdf450

yottta commented:
@Yantrio, @cam72cam. Thanks a lot for the reviews.



> Overall, this is a really good start! I think the main area I'd like to see expanded is the user documentation. Your example is quite comprehensive and also quite dense. I think introducing a few simpler examples, before going with the "Here's the a complex example showing how the simpler building blocks interplay" would be helpful.

For this, I reworked the RFC and moved all the information that was initially directly under the `Proposed solution`, into the `User Documentation` and made `User documentation` as a subchapter of `Proposed solution`.

Also, moved all the suggested errors to be shown in the `User documentation` and `Technical approach` is just stating what it needs to be done.

Gogotchuri commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,859 @@
+#### Write-only attributes
+This is a new concept that allows any existing `resource` to define attributes in its schema that can be only written without the ability to retrieve the value afterwards.
+
+By not being readable, this also means that an attribute configured by a provider this way, should not be written to the state or plan file either.
```
Correctness @grammarly
```suggestion
By not being readable, this also means that an attribute configured by a provider this way should not be written to the state or plan file either.
```

yottta replied:
Wanted to add the suggestions individually but the signoff of the commit is not working correctly from the GH UI, so I added all your suggestions in one commit: 4ee1ec72ed946b32d50637bb893c4c754d1b6e4e

Gogotchuri commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,859 @@
+
+A write-only attribute can accept an ephemeral or a non-ephemeral value, even though it's recommended to use ephemeral values for such attributes.
+
+Because these attributes are not written to the plan file, the update of a write-only attribute it's getting a little bit trickier.
```
Correctness And Clarity @grammarly
```suggestion
Because these attributes are not written to the plan file, updating a write-only attribute is getting a bit trickier.
```

yottta replied:
Wanted to add the suggestions individually but the signoff of the commit is not working correctly from the GH UI, so I added all your suggestions in one commit: 4ee1ec72ed946b32d50637bb893c4c754d1b6e4e

Gogotchuri commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,859 @@
+
+An `output` block from a root module cannot be marked as ephemeral.
+This limitation is natural since ephemeral outputs are meant to be skipped from the state file.
+Therefore, there is no usage of such a defined output block in a root module.
```
Correctness @grammarly
```suggestion
Therefore, there is no use for such a defined output block in a root module.
```

yottta replied:
Wanted to add the suggestions individually but the signoff of the commit is not working correctly from the GH UI, so I added all your suggestions in one commit: 4ee1ec72ed946b32d50637bb893c4c754d1b6e4e

Gogotchuri commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,859 @@
+│   on main.tf line 31, in resource "aws_secretsmanager_secret_version" "store_from_ephemeral_output":
+│   31:   secret_string = module.secret_management.secrets
+│
+│ "secret_string" cannot accept an ephemeral value because it is not a write-only attribute which means that will be written to the state.
```
Correctness @grammarly
```suggestion
│ "secret_string" cannot accept an ephemeral value because it is not a write-only attribute, meaning it will be written to the state.
```

Gogotchuri reviewed (approved):
Looks Great! Thanks @yottta 
I've used Grammarly, which flagged a few grammar issues that might improve the document's readability; I'm adding them as suggestions.

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,859 @@
+To do so, the flow of an ephemeral resource should look similar to the following:
+* Requests the information from the provider.
+* Is passing it into the evaluation context that will be used to evaluate expressions that are referencing it.
+  * Is not storing the value in plan or state file.
```
Wanted to add the suggestions individually but the signoff of the commit is not working correctly from the GH UI, so I added all your suggestions in one commit: 4ee1ec72ed946b32d50637bb893c4c754d1b6e4e

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,859 @@
+  * See the contexts above where an ephemeral resource can be referenced.
+* At the end, the ephemeral resource needs to be closed. OpenTofu should call `Close` on the provider for all opened ephemeral resources.
+
+Beside the attributes in a schema of an ephemeral resource, the block should support also the meta-arguments existing in OpenTofu:
```
Wanted to add the suggestions individually but the signoff of the commit is not working correctly from the GH UI, so I added all your suggestions in one commit: 4ee1ec72ed946b32d50637bb893c4c754d1b6e4e

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,859 @@
+  * ValidateConfig - validates the configuration provided for the ephemeral resource.
+
+Ephemeral resources lifecycle is similar with the data blocks:
+* Both basic implementations require the same methods (`Metadata` and `Schema`) while datasource is defining `Read` compared with the ephemeral resource that is defining `Open`. When talking about the basic functionality of the ephemeral resources, the `Open` method should behave similarly with the `Read` on a datasource, where it asks the provider for the data associated with that particular eph ...
```
Wanted to add the suggestions individually but the signoff of the commit is not working correctly from the GH UI, so I added all your suggestions in one commit: 4ee1ec72ed946b32d50637bb893c4c754d1b6e4e

yottta commented on rfc/20250317-ephemeral-resources.md:
````diff
@@ -0,0 +1,859 @@
+}
+```
+
+This function should work perfectly fine also with a non-ephemeral value.
````
Wanted to add the suggestions individually but the signoff of the commit is not working correctly from the GH UI, so I added all your suggestions in one commit: 4ee1ec72ed946b32d50637bb893c4c754d1b6e4e

yottta commented on rfc/20250317-ephemeral-resources.md:
```diff
@@ -0,0 +1,859 @@
+* write-only attributes
+* ephemeral resources
+* `provisioner` block
+* `connection` block
```
Added in 5c34b67c8d5e71d22fe4116af6bc9929ea2ee97d
